components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java

```diff
@@ -75,53 +75,45 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
             throws IdentityOAuth2Exception, IdentityOAuthAdminException, InvalidOAuthClientException {
 
-        String responseType = authzReqDTO.getResponseType();
-        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
-
-        if (!responseHandlers.containsKey(responseType)) {
-            log.warn("Unsupported Response Type : " + responseType +
-                    " provided  for user : " + authzReqDTO.getUser());
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNSUPPORTED_RESPONSE_TYPE,
-                    "Unsupported Response Type!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        OAuthAuthzReqMessageContext authzReqMsgCtx = getoAuthAuthzReqMessageContext(authzReqDTO);
+        ResponseTypeHandler authzHandler = responseHandlers.get(authzReqDTO.getResponseType());
+        OAuth2AuthorizeRespDTO authorizeRespDTO = validateAuthzRequest(authzReqDTO, authzReqMsgCtx, authzHandler);
+        if (foundErrorResponse(authorizeRespDTO)) {
             return authorizeRespDTO;
         }
+        try {
+	    // set the authorization request context to be used by downstream handlers. This is introduced as a fix for
+	    // IDENTITY-4111
+	    OAuth2Util.setAuthzRequestContext(authzReqMsgCtx);
+	    authorizeRespDTO = authzHandler.issue(authzReqMsgCtx);
+        } finally {
+            // clears authorization request context
+            OAuth2Util.clearAuthzRequestContext();
+        }
+        return authorizeRespDTO;
+    }
 
-        ResponseTypeHandler authzHandler = responseHandlers.get(responseType);
-        OAuthAuthzReqMessageContext authzReqMsgCtx = new OAuthAuthzReqMessageContext(authzReqDTO);
-
-        // loading the stored application data
-        OAuthAppDO oAuthAppDO = getAppInformation(authzReqDTO);
-        authzReqMsgCtx.addProperty("OAuthAppDO", oAuthAppDO);
-
-        // load the SP tenant domain from the OAuth App info
-        authzReqMsgCtx.getAuthorizationReqDTO().setTenantDomain(OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO));
-
-        boolean isAuthorizedClient = authzHandler.isAuthorizedClient(authzReqMsgCtx);
-
-        if (!isAuthorizedClient) {
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
-                    "The authenticated client is not authorized to use this authorization grant type");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+    private OAuth2AuthorizeRespDTO validateAuthzRequest(OAuth2AuthorizeReqDTO authzReqDTO,
+                                                        OAuthAuthzReqMessageContext authzReqMsgCtx,
+                                                        ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
+        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
+        if(!validateResponseType(authzReqDTO, authorizeRespDTO)) {
```

Comment:
Better to have a debug log

Reply:
Added debug logs within validateResponseType() method